File: src/common/config/loader.py
"""Load, validate, and freeze the YAML files under ``config/``.

Importable with only the standard library plus ``yaml`` -- no ``util.file``,
no domain modules -- so it can run before the rest of the application
(including the filesystem layer) is initialized.

Schema validation is a small hand-rolled walker (see ``_walk``) rather than a
full JSON Schema implementation. It supports exactly the keywords used by
``config/schema/*.schema.json`` today: ``type``, ``properties``,
``required``, ``additionalProperties``, ``items``, ``minItems``,
``maxItems``, ``uniqueItems``, ``minimum``, ``maximum``, ``exclusiveMinimum``,
``exclusiveMaximum``, ``const``, ``enum``, and ``pattern``. Any other keyword
in a schema is a startup error (``_UNSUPPORTED_KEYWORDS`` guard) rather than
a silently-unenforced constraint, so a schema author who reaches for
``oneOf``/``$ref``/``format`` finds out immediately instead of shipping a
schema that looks stricter than it is.

Precedence for locating the config root, highest first:
1. An explicit ``config_dir`` argument (typically sourced from a ``--config-dir``
   CLI flag).
2. The ``EDGEWARN_CONFIG_DIR`` environment variable.
3. A ``config/`` directory found by walking up from this file's location.
"""

from __future__ import annotations

import json
import os
import re
from pathlib import Path
from types import MappingProxyType
from typing import Any

import yaml

CONFIG_NAMES: tuple[str, ...] = (
    "runtime",
    "historical",
    "filesystem",
    "detection",
    "lineage",
    "integration",
    "scheduler",
    "api_index",
    "mrms_goes",
    "nexrad",
    "synoptic_rap",
    "wpc",
    "metar",
    "nws",
    "ewmrs_render",
    "ewmrs_rap_uint16",
    "ewmrs_pipeline",
    "api",
    "kalman",
)

_ENV_CONFIG_DIR = "EDGEWARN_CONFIG_DIR"

_config_cache: dict[tuple[str, str], Any] = {}
_provenance_cache: dict[tuple[str, str], dict[str, Any]] = {}


class ConfigError(Exception):
    """Raised for a missing config file, missing key, or schema violation."""

    def __init__(self, filename: str, dotted_path: str | None, message: str):
        self.filename = filename
        self.dotted_path = dotted_path
        self.message = message
        super().__init__(str(self))

    def __str__(self) -> str:
        if self.dotted_path:
            return f"{self.filename}: {self.dotted_path}: {self.message}"
        return f"{self.filename}: {self.message}"


def _find_config_root_by_walking_up() -> Path:
    here = Path(__file__).resolve()
    for candidate_dir in here.parents:
        config_dir = candidate_dir / "config"
        if (config_dir / "runtime.yaml").is_file():
            return config_dir
    raise ConfigError(
        "config/",
        None,
        f"could not locate a config/ directory containing runtime.yaml by "
        f"walking up from {here}",
    )


def config_root(cli_dir: str | os.PathLike[str] | None = None) -> Path:
    """Resolve the config root directory using CLI > env > repo-root precedence."""
    if cli_dir is not None:
        resolved = Path(cli_dir).resolve()
        if not (resolved / "runtime.yaml").is_file():
            raise ConfigError(
                str(resolved),
                None,
                "--config-dir does not contain runtime.yaml",
            )
        return resolved

    env_dir = os.environ.get(_ENV_CONFIG_DIR)
    if env_dir is not None:
        resolved = Path(env_dir).resolve()
        if not (resolved / "runtime.yaml").is_file():
            raise ConfigError(
                str(resolved),
                None,
                f"{_ENV_CONFIG_DIR} does not contain runtime.yaml",
            )
        return resolved

    return _find_config_root_by_walking_up()


def repo_root(cli_dir: str | os.PathLike[str] | None = None) -> Path:
    """The repository root, derived as the parent of the config root."""
    return config_root(cli_dir).parent


def _freeze(value: Any) -> Any:
    if isinstance(value, dict):
        return MappingProxyType({k: _freeze(v) for k, v in value.items()})
    if isinstance(value, list):
        return tuple(_freeze(v) for v in value)
    return value


def _dotted_path(path_parts: list[Any]) -> str:
    parts: list[str] = []
    for part in path_parts:
        if isinstance(part, int):
            parts[-1] = f"{parts[-1]}[{part}]"
        else:
            parts.append(str(part))
    return ".".join(parts)


_KNOWN_SCHEMA_KEYWORDS = frozenset({
    "$schema", "title", "description",
    "type", "properties", "required", "additionalProperties",
    "items", "minItems", "maxItems", "uniqueItems",
    "minimum", "maximum", "exclusiveMinimum", "exclusiveMaximum",
    "const", "enum", "pattern",
})


def _check_supported_keywords(schema_path: Path, node: Any, path: list[Any]) -> None:
    """Reject any schema keyword the hand-rolled walker doesn't implement.

    Without this, an author reaching for e.g. ``oneOf`` or ``$ref`` would get
    a schema that silently enforces nothing for that constraint instead of a
    startup error.
    """
    if not isinstance(node, dict):
        return
    unknown = sorted(set(node) - _KNOWN_SCHEMA_KEYWORDS)
    if unknown:
        raise ConfigError(
            str(schema_path),
            _dotted_path(path) or None,
            f"unsupported schema keyword(s) {unknown}",
        )
    for prop_name, prop_schema in node.get("properties", {}).items():
        _check_supported_keywords(schema_path, prop_schema, path + [prop_name])
    additional = node.get("additionalProperties")
    if isinstance(additional, dict):
        _check_supported_keywords(schema_path, additional, path + ["additionalProperties"])
    items = node.get("items")
    if isinstance(items, dict):
        _check_supported_keywords(schema_path, items, path + ["items"])


def _type_matches(value: Any, type_name: str) -> bool:
    if type_name == "object":
        return isinstance(value, dict)
    if type_name == "array":
        return isinstance(value, list)
    if type_name == "string":
        return isinstance(value, str)
    if type_name == "boolean":
        return isinstance(value, bool)
    if type_name == "integer":
        return isinstance(value, int) and not isinstance(value, bool)
    if type_name == "number":
        return isinstance(value, (int, float)) and not isinstance(value, bool)
    if type_name == "null":
        return value is None
    raise ConfigError("<schema>", None, f"unsupported schema type {type_name!r}")


def _has_duplicates(items: list[Any]) -> bool:
    seen: list[Any] = []
    for item in items:
        if item in seen:
            return True
        seen.append(item)
    return False


def _walk(schema: dict[str, Any], value: Any, path: list[Any], errors: list[tuple[list[Any], str]]) -> None:
    type_spec = schema.get("type")
    if type_spec is not None:
        type_names = [type_spec] if isinstance(type_spec, str) else list(type_spec)
        if not any(_type_matches(value, t) for t in type_names):
            errors.append((path, f"{value!r} is not of type {' or '.join(type_names)}"))
            return

    if "const" in schema and value != schema["const"]:
        errors.append((path, f"must equal {schema['const']!r}"))
    if "enum" in schema and value not in schema["enum"]:
        errors.append((path, f"must be one of {schema['enum']!r}"))

    if isinstance(value, dict):
        for key in schema.get("required", ()):
            if key not in value:
                errors.append((path + [key], "is a required property"))
        properties = schema.get("properties", {})
        additional = schema.get("additionalProperties", True)
        for key, sub_value in value.items():
            if key in properties:
                _walk(properties[key], sub_value, path + [key], errors)
            elif additional is False:
                errors.append((path + [key], "additional properties are not allowed"))
            elif isinstance(additional, dict):
                _walk(additional, sub_value, path + [key], errors)

    elif isinstance(value, list):
        min_items = schema.get("minItems")
        if min_items is not None and len(value) < min_items:
            errors.append((path, f"must have at least {min_items} item(s)"))
        max_items = schema.get("maxItems")
        if max_items is not None and len(value) > max_items:
            errors.append((path, f"must have at most {max_items} item(s)"))
        if schema.get("uniqueItems") and _has_duplicates(value):
            errors.append((path, "items must be unique"))
        item_schema = schema.get("items")
        if item_schema is not None:
            for index, item in enumerate(value):
                _walk(item_schema, item, path + [index], errors)

    elif isinstance(value, (int, float)) and not isinstance(value, bool):
        bound = schema.get("minimum")
        if bound is not None and value < bound:
            errors.append((path, f"must be >= {bound}"))
        bound = schema.get("maximum")
        if bound is not None and value > bound:
            errors.append((path, f"must be <= {bound}"))
        bound = schema.get("exclusiveMinimum")
        if bound is not None and value <= bound:
            errors.append((path, f"must be > {bound}"))
        bound = schema.get("exclusiveMaximum")
        if bound is not None and value >= bound:
            errors.append((path, f"must be < {bound}"))

    elif isinstance(value, str) and "pattern" in schema and re.search(schema["pattern"], value) is None:
        errors.append((path, f"does not match pattern {schema['pattern']!r}"))


def _validate(name: str, document: dict[str, Any], schema_path: Path) -> None:
    if not schema_path.is_file():
        raise ConfigError(str(schema_path), None, "schema file not found")

    with schema_path.open("r", encoding="utf-8") as fh:
        schema = json.load(fh)

    _check_supported_keywords(schema_path, schema, [])

    errors: list[tuple[list[Any], str]] = []
    _walk(schema, document, [], errors)
    if errors:
        first_path, first_message = min(errors, key=lambda e: (len(e[0]), [str(p) for p in e[0]]))
        raise ConfigError(f"{name}.yaml", _dotted_path(first_path) or None, first_message)


def reset_cache() -> None:
    """Clear memoized configs and provenance. Intended for tests."""
    _config_cache.clear()
    _provenance_cache.clear()


def load_config(name: str, *, config_dir: str | os.PathLike[str] | None = None) -> Any:
    """Load, schema-validate, and freeze ``config/<name>.yaml``.

    Memoized per resolved config root and name, so repeated calls (including
    across module re-execution under multiprocessing) are cheap and return
    the identical frozen object.
    """
    root = config_root(config_dir)
    cache_key = (str(root), name)
    if cache_key in _config_cache:
        return _config_cache[cache_key]

    yaml_path = root / f"{name}.yaml"
    if not yaml_path.is_file():
        raise ConfigError(f"{name}.yaml", None, f"file not found at {yaml_path}")

    with yaml_path.open("r", encoding="utf-8") as fh:
        document = yaml.safe_load(fh)

    if not isinstance(document, dict):
        raise ConfigError(f"{name}.yaml", None, "top-level document must be a mapping")

    schema_path = root / "schema" / f"{name}.schema.json"
    _validate(name, document, schema_path)

    frozen = _freeze(document)
    _config_cache[cache_key] = frozen
    _provenance_cache[cache_key] = {
        "path": str(yaml_path),
        "schema_version": document.get("schema_version"),
    }
    return frozen


def get_provenance(name: str, *, config_dir: str | os.PathLike[str] | None = None) -> dict[str, Any]:
    """Sanitized provenance (path + schema_version, no secrets) for a loaded config."""
    root = config_root(config_dir)
    cache_key = (str(root), name)
    if cache_key not in _provenance_cache:
        load_config(name, config_dir=config_dir)
    return dict(_provenance_cache[cache_key])

File: src/config/loader.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import yaml from 'js-yaml';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const ENV_CONFIG_DIR = 'EDGEWARN_CONFIG_DIR';

export const CONFIG_NAMES = Object.freeze([
  'runtime', 'historical', 'filesystem', 'detection', 'lineage',
  'integration', 'scheduler', 'api_index', 'mrms_goes', 'nexrad', 'synoptic_rap',
  'wpc', 'metar', 'nws', 'ewmrs_render', 'ewmrs_rap_uint16',
  'ewmrs_pipeline', 'api', 'kalman',
]);

export class ConfigError extends Error {
  constructor(filename, dottedPath, message) {
    super(dottedPath ? `${filename}: ${dottedPath}: ${message}` : `${filename}: ${message}`);
    this.name = 'ConfigError';
    this.filename = filename;
    this.dottedPath = dottedPath;
  }
}

const configCache = new Map();
const provenanceCache = new Map();

// Hand-rolled schema walker, mirroring src/common/config/loader.py's _walk.
// Supports exactly the keywords used by config/schema/*.schema.json today;
// anything else is a startup error rather than a silently-unenforced constraint.
const KNOWN_SCHEMA_KEYWORDS = new Set([
  '$schema', 'title', 'description',
  'type', 'properties', 'required', 'additionalProperties',
  'items', 'minItems', 'maxItems', 'uniqueItems',
  'minimum', 'maximum', 'exclusiveMinimum', 'exclusiveMaximum',
  'const', 'enum', 'pattern',
]);

function isPlainObject(value) {
  return value !== null && typeof value === 'object' && !Array.isArray(value);
}

function deepEqual(a, b) {
  if (a === b) return true;
  if (Array.isArray(a) && Array.isArray(b)) {
    return a.length === b.length && a.every((item, index) => deepEqual(item, b[index]));
  }
  if (isPlainObject(a) && isPlainObject(b)) {
    const aKeys = Object.keys(a);
    const bKeys = Object.keys(b);
    return aKeys.length === bKeys.length
      && aKeys.every((key) => Object.prototype.hasOwnProperty.call(b, key) && deepEqual(a[key], b[key]));
  }
  return false;
}

function hasDuplicates(items) {
  for (let i = 0; i < items.length; i += 1) {
    for (let j = i + 1; j < items.length; j += 1) {
      if (deepEqual(items[i], items[j])) return true;
    }
  }
  return false;
}

function typeMatches(value, typeName) {
  switch (typeName) {
    case 'object': return isPlainObject(value);
    case 'array': return Array.isArray(value);
    case 'string': return typeof value === 'string';
    case 'boolean': return typeof value === 'boolean';
    case 'integer': return typeof value === 'number' && Number.isInteger(value);
    case 'number': return typeof value === 'number';
    case 'null': return value === null;
    default: throw new ConfigError('<schema>', null, `unsupported schema type ${JSON.stringify(typeName)}`);
  }
}

function checkSupportedKeywords(schemaPath, node, pathParts) {
  if (!isPlainObject(node)) return;
  const unknown = Object.keys(node).filter((key) => !KNOWN_SCHEMA_KEYWORDS.has(key)).sort();
  if (unknown.length > 0) {
    throw new ConfigError(schemaPath, dottedPath(pathParts), `unsupported schema keyword(s) ${JSON.stringify(unknown)}`);
  }
  for (const [propName, propSchema] of Object.entries(node.properties || {})) {
    checkSupportedKeywords(schemaPath, propSchema, [...pathParts, propName]);
  }
  if (isPlainObject(node.additionalProperties)) {
    checkSupportedKeywords(schemaPath, node.additionalProperties, [...pathParts, 'additionalProperties']);
  }
  if (isPlainObject(node.items)) {
    checkSupportedKeywords(schemaPath, node.items, [...pathParts, 'items']);
  }
}

function walk(schema, value, pathParts, errors) {
  const typeSpec = schema.type;
  if (typeSpec !== undefined) {
    const typeNames = Array.isArray(typeSpec) ? typeSpec : [typeSpec];
    if (!typeNames.some((typeName) => typeMatches(value, typeName))) {
      errors.push([pathParts, `${JSON.stringify(value)} is not of type ${typeNames.join(' or ')}`]);
      return;
    }
  }

  if ('const' in schema && !deepEqual(value, schema.const)) {
    errors.push([pathParts, `must equal ${JSON.stringify(schema.const)}`]);
  }
  if ('enum' in schema && !schema.enum.some((option) => deepEqual(value, option))) {
    errors.push([pathParts, `must be one of ${JSON.stringify(schema.enum)}`]);
  }

  if (isPlainObject(value)) {
    for (const key of schema.required || []) {
      if (!Object.prototype.hasOwnProperty.call(value, key)) {
        errors.push([[...pathParts, key], 'is a required property']);
      }
    }
    const properties = schema.properties || {};
    const additional = 'additionalProperties' in schema ? schema.additionalProperties : true;
    for (const [key, subValue] of Object.entries(value)) {
      if (Object.prototype.hasOwnProperty.call(properties, key)) {
        walk(properties[key], subValue, [...pathParts, key], errors);
      } else if (additional === false) {
        errors.push([[...pathParts, key], 'additional properties are not allowed']);
      } else if (isPlainObject(additional)) {
        walk(additional, subValue, [...pathParts, key], errors);
      }
    }
  } else if (Array.isArray(value)) {
    if (schema.minItems !== undefined && value.length < schema.minItems) {
      errors.push([pathParts, `must have at least ${schema.minItems} item(s)`]);
    }
    if (schema.maxItems !== undefined && value.length > schema.maxItems) {
      errors.push([pathParts, `must have at most ${schema.maxItems} item(s)`]);
    }
    if (schema.uniqueItems && hasDuplicates(value)) {
      errors.push([pathParts, 'items must be unique']);
    }
    if (schema.items !== undefined) {
      value.forEach((item, index) => walk(schema.items, item, [...pathParts, index], errors));
    }
  } else if (typeof value === 'number') {
    if (schema.minimum !== undefined && value < schema.minimum) {
      errors.push([pathParts, `must be >= ${schema.minimum}`]);
    }
    if (schema.maximum !== undefined && value > schema.maximum) {
      errors.push([pathParts, `must be <= ${schema.maximum}`]);
    }
    if (schema.exclusiveMinimum !== undefined && value <= schema.exclusiveMinimum) {
      errors.push([pathParts, `must be > ${schema.exclusiveMinimum}`]);
    }
    if (schema.exclusiveMaximum !== undefined && value >= schema.exclusiveMaximum) {
      errors.push([pathParts, `must be < ${schema.exclusiveMaximum}`]);
    }
  } else if (typeof value === 'string' && schema.pattern !== undefined && !new RegExp(schema.pattern).test(value)) {
    errors.push([pathParts, `does not match pattern ${JSON.stringify(schema.pattern)}`]);
  }
}

function findConfigRootByWalkingUp() {
  let current = __dirname;
  while (true) {
    const parent = path.dirname(current);
    if (parent === current) break;
    current = parent;
    const configDir = path.join(current, 'config');
    if (fs.existsSync(path.join(configDir, 'runtime.yaml'))) {
      return configDir;
    }
  }
  throw new ConfigError('config/', null, `could not locate a config/ directory containing runtime.yaml by walking up from ${__dirname}`);
}

export function configRoot(cliDir = null) {
  if (cliDir) return path.resolve(cliDir);
  const envDir = process.env[ENV_CONFIG_DIR];
  if (envDir) return path.resolve(envDir);
  return findConfigRootByWalkingUp();
}

export function repoRoot(cliDir = null) {
  return path.dirname(configRoot(cliDir));
}

// Mirrors PATH_TOKENS and expand_path in src/common/config/loader.py. An
// allowlist rather than a scan for bare `<`/`>`, which would false-positive on
// comment lines and on synoptic_rap.yaml's named capture groups.
export const PATH_TOKENS = Object.freeze(['base_dir', 'gui_dir', 'src_dir']);

export function expandPath(template, roots, { filename, dottedPath }) {
  if (typeof template !== 'string') {
    throw new ConfigError(filename, dottedPath, `expected a path string, got ${JSON.stringify(template)}`);
  }
  // Checked before the prefix match so a Windows-style template is reported as
  // the separator problem it is, rather than as a malformed remainder.
  if (template.includes('\\')) {
    throw new ConfigError(filename, dottedPath, `${JSON.stringify(template)} must use '/' separators`);
  }
  // A NUL is never valid in a path. Node throws only once one reaches an fs call,
  // and expandPath returns a string that may be stored or logged well before that,
  // so it is rejected at the same point the Python side rejects it.
  if (template.includes('\0')) {
    throw new ConfigError(filename, dottedPath, 'path contains a NUL byte');
  }
  const match = /^<([a-z_]+)>\//.exec(template);
  if (match === null) {
    const expected = PATH_TOKENS.map((token) => `<${token}>/`).join(', ');
    throw new ConfigError(filename, dottedPath, `${JSON.stringify(template)} must begin with one of ${expected}`);
  }
  const token = match[1];
  if (!PATH_TOKENS.includes(token)) {
    throw new ConfigError(filename, dottedPath, `<${token}> is not an expandable path token`);
  }
  if (!Object.prototype.hasOwnProperty.call(roots, token)) {
    throw new ConfigError(filename, dottedPath, `<${token}> has no value in this context`);
  }
  const remainder = template.slice(match[0].length);
  const segments = remainder.split('/');
  if (remainder === '' || remainder.startsWith('/') || segments.includes('..')) {
    throw new ConfigError(filename, dottedPath, `${JSON.stringify(remainder)} is not a relative path below <${token}>`);
  }
  // Kept after the textual check rather than instead of it. This is the guard
  // the API config has always used, and it stays as the backstop that fails on
  // anything path.resolve normalizes out of the root; the check above is what
  // rejects a hostile template before any path is built from it. Unlike the
  // Python side, path.resolve does not follow symlinks, so this is defense in
  // depth rather than a second class of input.
  // An empty or relative root would make path.resolve fall back to process.cwd(),
  // reintroducing the working-directory dependence the mandatory token exists to
  // prevent -- and silently, since the result is still a plausible path.
  const givenRoot = roots[token];
  if (!givenRoot || !path.isAbsolute(givenRoot)) {
    throw new ConfigError(filename, dottedPath, `<${token}> must be an absolute directory, got ${JSON.stringify(givenRoot)}`);
  }
  const root = path.resolve(givenRoot);
  const resolved = path.resolve(root, remainder);
  const relative = path.relative(root, resolved);
  if (relative === '' || relative === '..' || relative.startsWith(`..${path.sep}`) || path.isAbsolute(relative)) {
    throw new ConfigError(filename, dottedPath, `${JSON.stringify(template)} resolves outside <${token}>`);
  }
  return resolved;
}

function deepFreeze(value) {
  if (Array.isArray(value)) {
    return Object.freeze(value.map(deepFreeze));
  }
  if (value !== null && typeof value === 'object') {
    const frozen = {};
    for (const [key, val] of Object.entries(value)) {
      frozen[key] = deepFreeze(val);
    }
    return Object.freeze(frozen);
  }
  return value;
}

function dottedPath(pathParts) {
  const parts = [];
  for (const part of pathParts) {
    if (typeof part === 'number') {
      parts[parts.length - 1] = `${parts[parts.length - 1]}[${part}]`;
    } else {
      parts.push(String(part));
    }
  }
  return parts.join('.') || null;
}

function validateDocument(name, document, schemaPath) {
  if (!fs.existsSync(schemaPath)) {
    throw new ConfigError(`${name}.yaml`, null, `missing schema file: ${schemaPath}`);
  }
  const schema = JSON.parse(fs.readFileSync(schemaPath, 'utf8'));
  checkSupportedKeywords(schemaPath, schema, []);

  const errors = [];
  walk(schema, document, [], errors);
  if (errors.length > 0) {
    errors.sort((a, b) => {
      const aKey = a[0].map(String);
      const bKey = b[0].map(String);
      if (aKey.length !== bKey.length) return aKey.length - bKey.length;
      return aKey.join(' ').localeCompare(bKey.join(' '));
    });
    const [firstPath, firstMessage] = errors[0];
    throw new ConfigError(`${name}.yaml`, dottedPath(firstPath), firstMessage);
  }
}

export function resetCache() {
  configCache.clear();
  provenanceCache.clear();
}

export function loadConfig(name, { configDir = null } = {}) {
  const root = configRoot(configDir);
  const cacheKey = `${root}::${name}`;
  if (configCache.has(cacheKey)) {
    return configCache.get(cacheKey);
  }

  const yamlPath = path.join(root, `${name}.yaml`);
  if (!fs.existsSync(yamlPath)) {
    throw new ConfigError(`${name}.yaml`, null, `missing config file: ${yamlPath}`);
  }

  const document = yaml.load(fs.readFileSync(yamlPath, 'utf8'));
  if (document === null || typeof document !== 'object' || Array.isArray(document)) {
    throw new ConfigError(`${name}.yaml`, null, 'expected a YAML mapping at the document root');
  }

  const schemaPath = path.join(root, 'schema', `${name}.schema.json`);
  validateDocument(name, document, schemaPath);

  const frozen = deepFreeze(document);
  configCache.set(cacheKey, frozen);
  provenanceCache.set(cacheKey, { path: yamlPath, schema_version: document.schema_version });
  return frozen;
}

export function getProvenance(name, { configDir = null } = {}) {
  const root = configRoot(configDir);
  const cacheKey = `${root}::${name}`;
  if (!provenanceCache.has(cacheKey)) {
    throw new ConfigError(`${name}.yaml`, null, 'config has not been loaded yet');
  }
  return provenanceCache.get(cacheKey);
}
